Source/WebKit/UIProcess/API/glib/WebKitImage.cpp

```diff
@@ -0,0 +1,476 @@
+    return image->priv->height;
+}
+
+/**
+ * webkit_image_get_stride:
+ * @image: a #WebKitImage
+ *
+ * Get the @image stride
+ *
+ * Returns: the image stride
+ *
+ * Since: 2.52
+ */
+guint webkit_image_get_stride(WebKitImage* image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), 0);
+    return image->priv->stride;
+}
+
+/**
+ * webkit_image_as_rgba_bytes:
+ * @image: a #WebKitImage
+ *
+ * Get the @image pixel data as an array of bytes.
+ *
+ * The pixel format for the returned byte buffer is 32-bit per pixel
+ * with 8-bit premultiplied alpha, in the preferred byte order for
+ * the architecture (typically ABGR8888 on little-endian hosts, and
+ * RGBA8888 on big-endian ones).
+ *
+ * Returns: (transfer none): a #GBytes
+ *
+ * Since: 2.52
+ */
+GBytes* webkit_image_as_rgba_bytes(WebKitImage* image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), nullptr);
+
+    if (!image->priv->bytes) {
+        ASSERT(image->priv->createBytesFn);
+        image->priv->bytes = image->priv->createBytesFn(image);
+    }
+    ASSERT(image->priv->bytes);
+
+    return image->priv->bytes.get();
+}
+
+#if PLATFORM(GTK)
+/**
+ * webkit_image_to_gdk_texture:
+ * @image: a #WebKitImage
+ *
+ * Get the @image as a #GdkTexture
+ *
+ * Returns: (transfer full): a #GdkTexture
+ *
+ * Since: 2.52
+ */
+GdkTexture* webkit_image_to_gdk_texture(WebKitImage *image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), nullptr);
+
+    auto* bytes = webkit_image_as_rgba_bytes(image);
+
+    return gdk_memory_texture_new(image->priv->width, image->priv->height, GDK_MEMORY_DEFAULT, bytes, image->priv->stride);
+}
+#endif
+
+static guint webkitImageHash(GIcon*);
+static gboolean webkitImageEqual(GIcon*, GIcon*);
+
+static GInputStream* webkitImageLoad(GLoadableIcon*, int, char**, GCancellable*, GError**);
+static void webkitImageLoadAsync(GLoadableIcon*, int, GCancellable*, GAsyncReadyCallback, gpointer);
+static GInputStream* webkitImageLoadFinish(GLoadableIcon*, GAsyncResult*, char**, GError**);
+
+static void webkit_image_gicon_interface_init(GIconIface* iface)
+{
+    iface->hash = webkitImageHash;
+    iface->equal = webkitImageEqual;
+}
+
+static void webkit_image_gloadable_icon_interface_init(GLoadableIconIface* iface)
+{
+    iface->load = webkitImageLoad;
+    iface->load_async = webkitImageLoadAsync;
+    iface->load_finish = webkitImageLoadFinish;
+}
+
+guint webkitImageHash(GIcon* icon)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(icon), 0);
+
+    auto* image = WEBKIT_IMAGE(icon);
+
+    Hasher hasher;
+
+    auto* bytes = webkit_image_as_rgba_bytes(image);
+    auto dataSpan = span(bytes);
+
+    gsize rowBytes = image->priv->width * RGBA8BytesPerPixel;
+
+    ASSERT(dataSpan.size() >= image->priv->stride * (image->priv->height - 1) + rowBytes);
+
+    for (int y = 0; y < image->priv->height; ++y) {
+        auto row = consumeSpan(dataSpan, image->priv->stride);
+        WTF::add(hasher, row.first(rowBytes));
```

Comment:
Why don't we just use g_bytes_hash?

Reply:
An earlier version used `g_bytes_hash`. The problem is that it will consider the entirety of the bytes, including any possible padding between the end of the row data and the stride.